src/lem/constrainer.rs

```diff
@@ -0,0 +1,344 @@
+                                &concrete_path,
+                                &alloc_has_match,
+                            )
+                            .map_err(|e| format!("{e}: failed to constrain `and`"))?;
+                            stack.push((op, Some(concrete_path_and_has_match), new_path_matchtag));
+                        } else {
+                            stack.push((op, Some(alloc_has_match), new_path_matchtag));
+                        }
+                    }
+
+                    // Now we need to enforce that at least one path was taken. We do that by constraining
+                    // that the sum of the previously collected `Boolean`s is one
+
+                    // If `branch_path_info` is Some, then we constrain using "concrete implies ..." logic
+                    if let Some(concrete_path) = branch_path_info {
+                        enforce_selector_with_premise(
+                            &mut cs.namespace(|| format!("{path}.enforce_selector_with_premise")),
+                            &concrete_path,
+                            &concrete_path_vec,
+                        )
+                        .map_err(|e| {
+                            format!("{e}: couldn't constrain `enforce_selector_with_premise`")
+                        })?;
+                    } else {
+                        // If `branch_path_info` is None, we just do regular constraining
+                        enforce_popcount_one(
+                            &mut cs.namespace(|| format!("{path}.enforce_popcount_one")),
+                            &concrete_path_vec[..],
+                        )
+                        .map_err(|e| format!("{e}: couldn't enforce `popcount_one`"))?;
+                    }
+                }
+                LEMOP::Seq(ops) => stack.extend(
+                    ops.iter()
+                        .rev()
+                        .map(|op| (op, branch_path_info.clone(), path.clone())),
+                ),
+                LEMOP::SetReturn(outputs) => {
+                    let is_concrete_path = Self::on_concrete_path(&branch_path_info)?;
+                    for (i, output) in outputs.iter().enumerate() {
+                        let Some(alloc_ptr_computed) = alloc_ptrs.get(output.name()) else {
+                            return Err(format!("Output {} not allocated", output.name()))
+                        };
+                        let z_ptr = {
+                            if is_concrete_path {
+                                let Some(ptr) = witness.ptrs.get(output.name()) else {
+                                    return Err(format!("Output {} not found in the witness", output.name()))
+                                };
+                                store.hydrate_ptr(ptr)?
+                            } else {
+                                ZPtr::dummy()
+                            }
+                        };
```

Comment:
```suggestion
                        let z_ptr = if is_concrete_path {
                            let ptr = witness.ptrs.get(output.name()).ok_or_else(|| {
                                anyhow!("Output {} not found in the witness", output.name())
                            })?;
                            store.hydrate_ptr(ptr)?
                        } else {
                            ZPtr::dummy()
                        };
```

Reply:
Can't apply suggestion due to conflict with previous suggestion